frontend/model/contracts/group.js

```diff
@@ -1229,6 +1236,48 @@ sbp('chelonia/defineContract', {
         if (profile) {
           Vue.set(profile, 'lastLoggedIn', meta.createdDate)
         }
+      },
+      sideEffect ({ meta, contractID }, { getters }) {
+        // Upon user's log-in, check for two possible notification emissions are executed here.
+        // 1. INCOME_DETAILS_OLD
+        // 2. NEAR_DISTRIBUTION_END
+        const { ourPayments, currentNotifications } = sbp('state/vuex/getters')
+        // NOTE: {} below is a temporary fix for a weird cypress bug that getters.groupProfiles[meta.username] is undefined when a user rejoins the same group after getting kicked out.
+        const { lastLoggedIn, incomeDetailsLastUpdatedDate, incomeDetailsType } = (getters.groupProfiles[meta.username] || {})
+        const myNotificationHas = checkFunc => currentNotifications.some(item => checkFunc(item))
+        const now = meta.createdDate
+
+        // 1. INCOME_DETAILS_OLD
+        //    - check if it's been over 6 months since the last time the user updated their income details.
+        if (incomeDetailsLastUpdatedDate &&
+          compareISOTimestamps(lastLoggedIn, incomeDetailsLastUpdatedDate) > 6 * MONTHS_MILLIS &&
+          // Also, check if the notification has already been sent for the recorded date. - prevents sending the same notification multiple times.
+          !myNotificationHas(item => item.type === 'INCOME_DETAILS_OLD' && item.data.lastUpdatedDate === incomeDetailsLastUpdatedDate)
+        ) {
+          sbp('gi.notifications/emit', 'INCOME_DETAILS_OLD', {
+            createdDate: now,
+            months: 6,
+            lastUpdatedDate: incomeDetailsLastUpdatedDate
+          })
+        }
+
+        // 2. NEAR_DISTRIBUTION_END
+        //    - check if it's within less than a week before the distribution period end and
+        //      let the pledging members know they have outstanding payment todo items if any.
+        const currentPeriod = getters.groupSettings.distributionDate
+        const nextPeriod = getters.periodAfterPeriod(currentPeriod)
+        if (incomeDetailsType === 'pledgeAmount' &&
+          comparePeriodStamps(nextPeriod, now) < DAYS_MILLIS * 7 &&
+          (ourPayments && ourPayments.todo?.length > 0) &&
+          // Also, check if the notification has already been sent for the target period. - prevents sending the same notification multiple times.
+          !myNotificationHas(item => item.type === 'NEAR_DISTRIBUTION_END' && item.data.period === currentPeriod)
+        ) {
+          sbp('gi.notifications/emit', 'NEAR_DISTRIBUTION_END', {
+            createdDate: now,
+            groupID: contractID,
+            period: currentPeriod
+          })
+        }
```

Comment:
This is a lot of logic to add to the contracts. Ideally - the contract code should be as simple as possible to avoid bugs and to keep the contract file small, as well as limiting the number of messages we send to the contracts (to not unnecessarily fill up servers with junk).

What this PR is trying to do is 2 things, I think:

1. Emit a notification when we've entered a new distribution period. This doesn't need to involve any contract code.
2. Emit a notification when it's 1 week before the end a distribution period.

AFAICT, neither of these things need to be done in the contract code. The logic for it can be done entirely locally.

And something that I've noticed is that this type of code currently exists in different places in the project. There is frequently a need to notify the user about something, and the checks for that are all over the place. This is a problem - because periodic checks to notify users when certain events are happening or about to happen should all be organized neatly in a single place in the app. There should be a single go-to file for adding such checks, and the logic for adding these checks should look and behave the same way.

For example, there is a function in `main.js` called `periodicStateChecks` that also has this problem. It's currently used to notify users about expiring proposals, and it also has the problem that it sends messages to the contracts (to mark globally sync'd state to set a boolean that says the notification was done, to prevent multiple notifications).

That check, and these checks, would be better done in a single file, and locally, without touching the contracts. And we should take this opportunity to do that correctly now, to prevent this problem from exploding into more spaghetti code down the line.

Here's what I propose:

Create a file that contains an array of objects that have this format:

```js
const every1MinQueue = [
  {
    notificationStateKey: '', // string or function that returns a string
                              // used to store a boolean in the Vuex
                              // rootState.periodicNotifications
    notificationCondition (): boolean {
      // returns a boolean that determins whether or not to
      // emit a notification
      return goodTimeToEmtiNotification() && !rootState.periodicNotifications[this.notificationStateKey]
    },
    notificationEmit () {
      // logic to emit a notifications
    },
    shouldClearNotificationStateKey (): boolean {
      // returns true to delete the
      // rootState.periodicNotifications[this.notificationStateKey]
      return enteredANewPeriod()
    }
  },
  // ... more periodic notification objects
]
const every5MinQueue = [
  // ... same format as above
]
const every30minQueue = [
  // ... same format as above
]

// to avoid using a lot of CPU, use the appropriate queue for tasks
// we don't want to perform intensive operations every minute unless
// absolutely necessary

setInterval(runCronQueue(every1MinQueue), 1 * MINS_MILLIS)
setInterval(runCronQueue(every5MinQueue), 5 * MINS_MILLIS)
setInterval(runCronQueue(every30MinQueue), 30 * MINS_MILLIS)
```

Then move all of the logic from this PR, and the logic from main's `periodicStateChecks` into that file, and delete the old way of doing things.

Reply:
Hi @taoeffect ,
Got a couple of questions regarding the above.

**Q1.** If I understood it right(please correct me if not), we would want to keep the `notificationStateKey` here in the root vuex state as boolean like below,

```js
{
  ...
  periodicNotifications: {
    key1: true,
    key2: false,
    ...
  }
}
```
What does each key being either `true` or `false` mean? are we keeping them to be able to toggle those notifications on/off when we want to in the app?

**Q2.** It's about how to compose the logic of `runCronQueue` function here. I have roughly written it down below based on what I understood from the comment. 

```js
function runCronQueue (queue) {
  const rootState = sbp('state/vuex/state')

  for (const entry of queue) {
    if (rootState.periodicNotifications[entry.notificationStateKey] && entry.notificationCondition()) {
      entry.notificationEmit()

      if (entry.shouldClearNotificationStateKey()) {
        Vue.delete(rootState.periodicNotifications, entry.notificationKey)
      }
    }
  }
}

```

So on each interval, we go through the queue and take below steps for each entry.
**1)** check if the `notificationStateKey` is `true` in the rootState and also if `notificationCondition()` is truthy.
**2)** If conditions in the step **1)** are met, execute `notificationEmit()` logic
**3)** As the post action of **2)**, check if `shouldClearNotificationStateKey()` returns `true`, we clear the `notificationStateKey` from the root state so that the current entry is not executed from the next interval.

 is the above correct?
 
 Thanks,

Reply:
> What does each key being either true or false mean? are we keeping them to be able to toggle those notifications on/off when we want to in the app?

They indicate whether or not the notification has already been fired. You can make this clearer by naming the rootState key something like `periodicNotificationsAlreadyFiredMap` instead of `periodicNotifications`. Using long descriptive variable names is a perfectly good stylistic way of coding that helps make the code more readable and avoids having to write a bunch of comments.

> Q2. It's about how to compose the logic of runCronQueue function here. I have roughly written it down below based on what I understood from the comment.

The first condition of the `if` that checks `notificationStateKey` needs to be negated, since we don't want to fire the same notification multiple times.

The other nested `if` condition should be outside of the first `if`, since if you keep it inside of the `if`, it won't get fired (because we do into the `if` only when a notification *hasn't* been fired - and the second condition — for clearing notifications — only happen after they've been fired)

Reply:
> because we do into the if only when a notification hasn't been fired

Sorry, Could you help with what this mean?

Reply:
*oops, I edited the comment to fix a typo. I meant "go" instead of "do"

Reply:
Thanks for clarification @taoeffect ,
Just to make sure my understanding is correct before moving on, Does below looks good?

```js
function runCronQueue (queue) {
  const firedMap = sbp('state/vuex/state').periodicNotificationsAlreadyFiredMap

  for (const entry of queue) {
    if (!(entry.notificationStateKey in firedMap)) { continue } // if the key has already been cleared. ignore this entry.

    if (firedMap[entry.notificationStateKey] && entry.shouldClearNotificationStateKey()) {
      // if the notification has already been fired, check & clear the key
      Vue.delete(rootState.periodicNotifications, entry.notificationKey)
    } else if (entry.notificationCondition()) {
      entry.notificationEmit()
      firedMap[entry.notificationStateKey] = true
    }
  }
}
```

Reply:
@SebinSong let's take this to Slack :)